### ISSUE TITLE
High CPU Usage for throbber loading animation on windows with many tabs 

### ISSUE THREAD

**Issue by Lej77** [piroor/treestyletab#1703]
## Short description
I noticed that the throbber loading animation caused hight CPU usage when I had many tabs open. I had only 1 loading tab that used the animation and about 300 other open tabs and my CPU usage went to about 30% for an infinitely loading dummy tab that I used for testing.

## Steps to reproduce

 1. Start Firefox with clean profile.
 2. Install TST.
 3. Download  ["iperf-2.0.9-win64"](https://iperf.fr/iperf-download.php) and run the program with the command `iperf.exe -s -p 8888`.
 4. Open an infinitely loading tab by entering the following URL: [localhost:8888](url).
 5. Observe Firefox CPU usage.
 6. Stop the loading tab.
 7. Open about 100 new tabs (hold `ctrl+t`) and wait for them to finish loading.
 8. Open an infinitely loading tab by entering the following URL: [localhost:8888](url).
 9. Observe high CPU usage for Firefox.

## Expected result
Throbber loading animation' CPU usage should not vary based on number of opened tabs.

## Actual result
Throbber loading animation' CPU usage does vary based on number of opened tabs.

## Environment

 * Platform (OS): Windows 10
 * Version of Firefox: 57.0.3 also on 59
 * Version (or revision) of Tree Style Tab: 2.4.3

**Comment by Lej77** [comment]
This seems to be caused by a Firefox bug. When the animated throbber element is not drawn by Firefox due to optimizations it causes High CPU usage. 

Currently this will always happen since the master throbber element used for animation synchronization is placed offscreen and Firefox will therefore not draw it. To fix this we can place it on screen and then place another element over it to hide it. (Hiding it with `opacity: 0;` or `display: none;` would not work since that would also stop Firefox from drawing it.)

Here is some CSS code that fixes this (it moves the master throbber element located in the `dummy-tabs` section to on-screen area and then covers it using the `extra-background` element):
```CSS
/* TST 2.4.3 HotFix for #1703 - Fix for Performance Issues due to Hidden Master Loading Throbber */
#dummy-tabs {
  top: 0px;
}
#extra-background {
  background: var(--tabbar-bg);
}
```

------------------------------

This issue can also occur when a loading tab is located outside of the visible area of the sidebar. That is to say when it the sidebar is scrolled so that a tab is not visible. The issue doesn't occur immediately when a tab gets scrolled outside of the visible area since Firefox seems to keep drawing parts outside of the visible area as well. From my testing it seems that Firefox keeps drawing in an area with about 1.5 times the visible areas height outside the visible area. 

For example (location: [area top]-[area bottom]):
// Unloaded area (0-500px)
/ Cached area (500-1000px)
/ Cached area (1000-1500px)
/ Cached area (1500-2000px)
| Visible area  (2000-2500px)
\ Cached area (2500-3000px)
\ Cached area (3000-3500px)
\ Cached area (3500-4000px)
\\ Unloaded area (4000-4500px)

A possible fix would be to stop the loading animation for tabs if they are too far away from the visible area of the sidebar.

**Comment by birtles** [comment]
Yes this appears to be [Firefox bug 1421507](https://bugzilla.mozilla.org/show_bug.cgi?id=1421507). I hope we can fix that bug soon.

**Comment by sumyungguy** [comment]
I put this in the "extra style rules for sidebar contents" preference:
```css
.throbber { 
  display: none !important;
}
```
It just removes the loading animation, at least until this issue can be fixed. I found it speeds up page loading very noticeably.

**Comment by piroor** [comment]
Thank you for detailed researching. Should I fix this by TST side? or, keep TST as is and wait for fixes by Firefox-side?

**Comment by Lej77** [comment]
@piroor You should probably fix the hidden master throbber since it is such an easy fix. The second issue with tabs that are scrolled out of view probably occurs much less often so maybe only fix it if it is easy and you have the time?

**Comment by piroor** [comment]
OK, I've introduced new background element to cover dummy elements by the commit 4eaf990. `#extra-background` was just for mixing of system colors, so it is now migrated to `#background::after`.
